tasks/sudodoki/2016-05-07-task-7/sets-operations.js

```diff
@@ -0,0 +1,84 @@
+// Задание 2: Эффективно реализовать множественные операции объединение, пересечение и
+// разница с использованием отсортированных списков
+'use strict';
+const printList = (list) => {
+  var head = list;
+  var consoleArgs = [''];
+  while (head) {
+      consoleArgs  = [consoleArgs[0].concat('%c %s %c⤏')].concat(consoleArgs.slice(1)).concat('background: #888; color: #222', head.el, 'background: #fff; color: #888');
+      head = head.next;
+  }
+  consoleArgs = [consoleArgs[0].concat('%c%s')].concat(consoleArgs.slice(1)).concat('background: #888; color: #222', 'NULL')
+  console.log.apply(console, consoleArgs);
+}
+
+const addToList = (list, el) => {
+  var tail = list;
+  while (tail.next) {
+    tail = tail.next
+  }
+  // handle list start
+  if (tail.hasOwnProperty('next')) {
+    tail.next = {el, next: undefined}
+  } else {
+    Object.assign(list, {el, next: undefined})
+  }
+}
+
+const buildList = (arr) => {
+  var list = {};
+  arr.forEach(el => addToList(list, el));
+  return list;
+}
+
+const buildSortedList = (arr) => buildList(JSON.parse(JSON.stringify(arr)).sort())
+
+const union = (left, right) => {
+  const res = {};
+  let [headLeft, headRight] = left.el > right.el ? [right, left] : [left, right];
+  while (headLeft) {
+    addToList(res, headLeft.el);
+    while (headRight && (headRight.el >= headLeft.el) && (headRight.el < (headLeft.next && headLeft.next.el || Infinity))) {
```

Comment:
moving in second list while items are bigger that currently taken element from first list but until items are smaller than next element in the first list. For last element there's no next element, so we take everything left (because it would be less than Inifnity)
```
1-------4
--2--3--5
```
So, after 1 we are taking 2, 3, before taking 4. After that we will take 5 because 4.next == undefined & 5 < Inifinity